src/components/common/needsShutdown.jsx

```diff
@@ -0,0 +1,168 @@
+            (vm.vcpus.max !== vm.inactiveXML.vcpus.max) ||
+            (vm.cpu.sockets !== vm.inactiveXML.cpu.sockets) ||
+            (vm.cpu.threads !== vm.inactiveXML.cpu.threads) ||
+            (vm.cpu.cores !== vm.inactiveXML.cpu.cores));
+}
+
+export function needsShutdownCpuModel(vm) {
+    /* The live xml shows what host-model expanded to when started
+     * This is important since the expansion varies depending on the host and so needs to be tracked across migration
+     */
+    if (vm.inactiveXML.cpu.mode === 'host-model')
+        return !(vm.cpu.mode == 'host-model' || vm.cpu.model === vm.capabilities.cpuHostModel);
+    if (vm.inactiveXML.cpu.mode === 'host-passthrough')
+        return vm.cpu.mode !== 'host-passthrough';
+    if (vm.inactiveXML.cpu.mode === 'custom')
+        return vm.cpu.mode !== 'custom' || vm.cpu.model !== vm.inactiveXML.cpu.model;
+
+    return false;
+}
+
+export function needsShutdownWatchdog(vm) {
+    return vm.persistent && vm.state === "running" && vm.inactiveXML.watchdog.action !== vm.watchdog.action;
+}
+
+export function getDevicesRequiringShutdown(vm) {
+    if (!vm.persistent)
+        return [];
+
+    const devices = [];
+
+    // DISKS
+    for (const target in vm.disks) {
+        if (needsShutdownDiskAccess(vm, target)) {
+            devices.push(_("Disk"));
+            break;
+        }
+    }
+
+    // INTERFACES
+    for (const iface of vm.interfaces) {
+        if (needsShutdownIfaceType(vm, iface) ||
+            needsShutdownIfaceModel(vm, iface) ||
+            needsShutdownIfaceSource(vm, iface)) {
+            devices.push(_("Network interface"));
+            break;
+        }
+    }
+
+    // VCPU
+    if (needsShutdownVcpu(vm))
+        devices.push(_("vCPUs"));
+
+    // CPU
+    if (needsShutdownCpuModel(vm))
+        devices.push(_("CPU"));
+
+    // Watchdog
+    if (needsShutdownWatchdog(vm))
+        devices.push(_("Watchdog"));
+
+    return devices;
+}
+
+export const NeedsShutdownTooltip = ({ iconId, tooltipId }) => {
+    return (
+        <Tooltip id={tooltipId} content={_("Changes will take effect after shutting down the VM")}>
+            <PendingIcon color="orange" id={iconId} />
+        </Tooltip>
+    );
+};
+
+NeedsShutdownTooltip.propTypes = {
+    iconId: PropTypes.string.isRequired,
+    tooltipId: PropTypes.string.isRequired,
+};
+
+export const NeedsShutdownAlert = ({ idPrefix }) =>
+    <Alert isInline variant='warning' id={`${idPrefix}-idle-message`} title={_("Changes will take effect after shutting down the VM")} />;
+
+NeedsShutdownAlert.propTypes = {
+    idPrefix: PropTypes.string.isRequired,
+};
+
+export const VmNeedsShutdown = ({ vm }) => {
+    const devices = getDevicesRequiringShutdown(vm);
+
+    if (devices.length === 0)
+        return;
+
+    const body = (
+        <>
+            <List isPlain isBordered>
+                { devices.map((device, idx) => {
+                    return (
+                        <ListItem key={idx}>
```

Comment:
Why don't you reuse the 'device' variable for key? Also it's suboptimal to use only indexes for keys.